### PR TITLE
Make cube.exponentiate lazy

### PR DESCRIPTION
**Pull request by dkillick**
Changes the last line in `iris.analysis.maths._math_op_common` to use lazy data. This means `iris.analysis.maths.exponentiate` (providing `cube ** exponent` functionality) is now lazy.

As I was in the space, I've also added some tests for `iris.analysis.maths.exponentiate`. These do not follow the pattern of the other test modules in the directory as we do not make use of an `operator` attribute for cube exponentiation.

### CODE REVIEW

File: lib/iris/tests/unit/analysis/maths/test_exponentiate.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        result = exponentiate(cube, self.exponent)
+        self.assertMaskedArrayEqual(result.data, expected)
+
+    @tests.skip_data
+    def test_lazy_data(self):
+        # Confirm that the cube's lazy data is preserved through the operation.
+        test_data = tests.get_data_path(('PP', 'simple_pp', 'global.pp'))
+        cube = iris.load_cube(test_data)
+        exponentiate(cube, self.exponent, in_place=True)
+        self.assertTrue(cube.has_lazy_data())
```
Comment (pelson) [inline lib/iris/tests/unit/analysis/maths/test_exponentiate.py]: You missed the crucial step here. If you checked the data, I think you will find that is isn't correct.

File: lib/iris/analysis/maths.py [hunk 2]
```diff
@@ -658,7 +658,7 @@ def _math_op_common(cube, operation_function, new_unit, in_place=False):
             operation_function(new_cube._my_data, out=new_cube._my_data)
         except TypeError:
             # Non ufunc function
-            operation_function(new_cube.data)
+            operation_function(new_cube._my_data)
```
Comment (pelson) [inline lib/iris/analysis/maths.py]: This must surely write into ```new_cube._my_data = ```

File: lib/iris/analysis/maths.py [hunk 3]
```diff
@@ -384,8 +384,8 @@ def exponentiate(cube, exponent, in_place=False):
     """
     _assert_is_cube(cube)
 
-    def power(data, out=None):
-        return np.power(data, exponent, out)
+    def power(data):
+        return np.power(data, exponent)
```
Comment (pelson) [inline lib/iris/analysis/maths.py]: Did you mean biggus here too?

Comment (dkillick) [inline lib/iris/analysis/maths.py]: Actually no - the tests were suggesting that I had preserved lazy data when exponentiating a cube, which was the full extent of the change I was aiming for here.

However I seem to have hit a bit of the worst of both worlds with this closure nonetheless: no use of biggus and no inplace possibility.

File: lib/iris/analysis/maths.py [hunk 4]
```diff
@@ -658,7 +658,7 @@ def _math_op_common(cube, operation_function, new_unit, in_place=False):
             operation_function(new_cube._my_data, out=new_cube._my_data)
         except TypeError:
```
Comment (pelson) [inline lib/iris/analysis/maths.py]: Proposal: remove the try except. The new meaning of "inplace" is "operate on the cube provided" not necessarily "operate on the data provided".

Therefore this should become:

```
new_cube._my_data = operation_function(new_cube._my_data)
```

**Comment by dkillick** [comment]
Review actions. I've simplified `maths._math_op_common` and got `maths.exponentiate` using `biggus.power`. I've also updated the tests with more coverage. Note that the masked array test is fixed by [biggus #167](https://github.com/SciTools/biggus/pull/167).

File: lib/iris/tests/unit/analysis/maths/test_exponentiate.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    def test_masked(self):
+        cube = self.cube.copy()
+        mask = cube.data % 3 == 0
+        masked_data = np.ma.masked_array(cube.data, mask)
+        cube.data = masked_data
+        expected = masked_data ** 2
+        result = exponentiate(cube, self.exponent)
+        self.assertMaskedArrayEqual(result.data, expected)
+
+    @tests.skip_data
```
Comment (pelson) [inline lib/iris/tests/unit/analysis/maths/test_exponentiate.py]: We don't really need to use real data here. We can just use the ```lazy_array``` method to get lazy data on the cube.

**Comment by pelson** [comment]
I think this is getting there. Is there documentation which needs updating? Is there in_place docs somewhere which describe the behaviour?

**Comment by dkillick** [comment]
> Is there documentation which [sic] needs updating?

I've taken a look; there is nothing in the cube mathematics section of the userguide about performing operations inplace. In the [reference guide](http://scitools.org.uk/iris/docs/latest/iris/iris/analysis/maths.html#iris.analysis.maths.exponentiate) almost all of the mathematical operators have the `in_place` kwarg described as follows:

```
    in_place:
        Whether to create a new Cube, or alter the given “cube”.
```

It's pretty terse, which actually means it is still correct. I could update it:

```
    in_place:
        Whether to create a new Cube as a copy of the given "cube" and alter its data to be
        the result of applying the mathematical operation, or alter the data of the given
        “cube” to be the result of applying the mathematical operation.
```

But that's increased the length of the description, which has not necessarily helped its cause.

**Comment by pelson** [comment]
> But that's increased the length of the description, which has not necessarily helped its cause.

Agree, that doesn't really help. Maybe just a what's new item then.

**Comment by pp-mo** [comment]
>  Is there documentation which needs updating?

++ whatsnew ??

**Comment by pelson** [comment]
@pp-mo - is this lined for 1.9?

**Comment by pp-mo** [comment]
> @pelson pp-mo - is this lined for 1.9?

Not unless you want to push for it.
We will have lazy basic arithmetic (+-*/) covered in the 1.9 whatsnew, though

**Comment by dkillick** [comment]
Rebase. Wonder whether the Biggus updates between when I opened this and now will have helped move this along?

**Comment by corinnebosley** [comment]
@dkillick Is there anything that can be done with this to get it moving again? The ticket is a year old.  That's a bit ridiculous.  Let me know if I can do something about this or bother someone who can.

**Comment by dkillick** [comment]
Another six months, another rebase...

**Comment by corinnebosley** [comment]
@dkillick Well that sucks.

**Comment by corinnebosley** [comment]
@dkillick 'The build has been terminated'.... Must be because it's Friday afternoon.  Travis is probably at the pub by now.

File: lib/iris/tests/unit/analysis/maths/test_exponentiate.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
+# (C) British Crown Copyright 2015, Met Office
```
Comment (dkillick) [inline lib/iris/tests/unit/analysis/maths/test_exponentiate.py]: That is far from being our only problem I suspect... Is it really complaining about the license header??

Comment (QuLogic) [inline lib/iris/tests/unit/analysis/maths/test_exponentiate.py]: So far, but the tests didn't complete, so...

**Comment by QuLogic** [comment]
I'd say see #2229.

**Comment by dkillick** [comment]
@QuLogic this (and #2229) may be helped by #2240. I'll restart the build to see if that's the case.

**Comment by dkillick** [comment]
Except, of course, we can't make use of the test runner changes in #2240 until the mergeback (#2236) is, well, merged back...

**Comment by marqh** [comment]
> Except, of course, we can't make use of the test runner changes in #2240 until the mergeback (#2236) is, well, merged back...

@dkillick this is now merged, so if you would be knid enough to re-rebase, that would be helpful

**Comment by dkillick** [comment]
> if you would be knid enough to re-rebase

Done (and mercifully painless!) - thanks @marqh

**Comment by marqh** [comment]
there is now an explicit text failure that needs addressing:
`FAIL: test_square_root (iris.tests.test_basic_maths.TestExponentiate)`

**Comment by marqh** [comment]
Hi @dkillick 

please could you review the test failures and see if this PR can be brought up to speed?

If successful, please consider targeting the 1.12.x branch for 1.12 release

thank you
mark

**Comment by pelson** [comment]
This PR is probably quite out of date with the recent dask work for v2. It might be worth checking to see if we have already go exp to be lazy at this point...

**Comment by pelson** [comment]
Update: it's ambiguous as to whether it is lazy, but there is certainly *scope* for it to be lazy with the current implementation. https://github.com/SciTools/iris/blob/master/lib/iris/analysis/maths.py#L542 has the line:

``return operator.pow(data, exponent)``

Which presumably means that if data is a dask array, and it has lazy exp, then iris does lazy exp... 🎉

**Pull request closed without merging**